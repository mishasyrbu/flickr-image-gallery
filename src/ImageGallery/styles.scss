.image-gallery {
    height:  100%;
    display: flex;
    flex-direction: column;
    align-items: center;
    padding-top: 20px;
    background-color: lightblue;

    > .image-list {
        display: grid;
        grid-template-columns: auto auto auto auto;
        grid-gap: 20px;
        overflow: auto;
        padding: 0 10%;
        margin-top: 20px;

        @media (max-width: 950px) {
            grid-template-columns: auto auto auto;
        }

        @media (max-width: 720px) {
            grid-template-columns: auto auto;
        }

        @media (max-width: 480px) {
            grid-template-columns: auto;
        }

        > img {
            width: 100%;
            height: 140px;
            object-fit: cover;
        }
    }
}
