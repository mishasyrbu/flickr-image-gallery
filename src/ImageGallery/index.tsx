import React, { SyntheticEvent } from 'react';
import { debounce } from 'lodash';

import SearchInput from '../components/SearchInput';
import ImageItem from '../components/ImageItem';
import { getImageUrl } from '../utils';
import { loadPhotosApi } from '../services';
import './styles.scss';

export interface ImageGalleryProps {
  
}

export interface PhotoObjet {
  id: string;
  owner: string;
  secret: string;
  server: string;
  farm: number;
  title: string;
}

export interface ImageGalleryState {
  searchQuery: string;
  photo: Array<PhotoObjet>;
  pages: number;
  perpage: number;
  total: string;
  page: number;
  loading: boolean;
}

class ImageGallery extends React.Component<ImageGalleryProps, ImageGalleryState> {
  constructor(props: ImageGalleryProps) {
    super(props);

    this.state = {
      searchQuery: '',
      photo: [],
      pages: 0,
      page: 1,
      perpage: 0,
      total: '0',
      loading: true,
    };

    this.onScrollList = debounce(this.onScrollList, 100);
    this.loadPhotos = debounce(this.loadPhotos, 500);
  }

  onScrollList = (scrollTop: number, offsetHeight: number, scrollHeight: number) => {
    const { page, pages } = this.state;
    if (scrollTop + offsetHeight >= scrollHeight - 350 && page < pages) {
      this.setState({ loading: true }, () => this.loadPhotos(true));
    }
  };

  handleOnScrollEvent = (event: SyntheticEvent<HTMLDivElement>) => {
    const { scrollTop, offsetHeight, scrollHeight } = event.currentTarget;
    
    this.onScrollList(scrollTop, offsetHeight, scrollHeight);
  };

  handleSearchChange = (searchQuery: string) => {
    this.setState({ searchQuery, page: 1, loading: true }, this.loadPhotos);
  };

  /**
   * Call load photos API and save result in state
   * 
   * scrollLoading {boolean} - flag indicates if we are in infinite scrolling state
   */
  loadPhotos = (scrollLoading = false) => {
    const { searchQuery, page, pages } = this.state;
    let nextPage = page;

    if (scrollLoading && page < pages) {
      nextPage++;
    }

    loadPhotosApi(searchQuery, nextPage).then((photos) => {
      if (scrollLoading) {
        this.setState((state) => ({ ...photos, photo: [...state.photo, ...photos.photo], loading: false }));
      } else {
        this.setState({ ...photos, loading: false });
      }
    });
  };

  render() {
    const { searchQuery, photo, total, loading } = this.state;

    return (
      <div className="image-gallery">
        {loading && <div className="loading-overlay">Loading...</div>}
        <SearchInput placeholder="Search..." value={searchQuery} onChange={this.handleSearchChange} />
        <div className="results-count">Found: {total}</div>
        <div className="image-list" onScroll={this.handleOnScrollEvent}>
          {photo.map(({ id, secret, server, farm, title }) => (
            <ImageItem key={`${farm}_${id}`} src={getImageUrl(farm, server, id, secret)} alt={title} />
          ))}
        </div>
      </div>
    );
  }

  componentDidMount() {
    this.loadPhotos();
  }
}

export default ImageGallery;
