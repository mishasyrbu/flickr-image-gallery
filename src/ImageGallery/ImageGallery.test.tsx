import React from 'react';
import Enzyme, { mount, ReactWrapper } from 'enzyme';
import { create, ReactTestRenderer } from 'react-test-renderer';

import ImageGallery, { ImageGalleryState, ImageGalleryProps } from './';

let wrapper: ReactWrapper<ImageGalleryProps, ImageGalleryState>;
let snapshot: ReactTestRenderer;

beforeEach(() => {
    jest.mock('../services', () => ({
        loadPhotosApi: jest.fn(),
    }));

    const imageGallery = <ImageGallery />;

    wrapper = mount(imageGallery);
    snapshot = create(imageGallery);
});

describe('<ImageGallery />', () => {
  test('it matches the snapshot', () => {
    expect(snapshot.toJSON()).toMatchSnapshot();
  });

  it('it should load photos on search query change event', () => {
    const spyLoadPhotos = jest.spyOn(wrapper.instance() as ImageGallery, 'loadPhotos');
    wrapper.update();
    wrapper.find('.search-input').simulate('change');

    expect(spyLoadPhotos).toHaveBeenCalled();
  });

  it('it should handle on scroll event', () => {
    const componentInstance = wrapper.instance() as ImageGallery;
    const spyOnScrollList = jest.spyOn(componentInstance, 'onScrollList');
    wrapper.update();
    wrapper.find('.image-list').simulate('scroll');

    expect(spyOnScrollList).toHaveBeenCalled();
  });

  it('it should update state on search query change event', () => {
    const componentInstance = wrapper.instance() as ImageGallery;
    const spyLoadPhotos = jest.spyOn(componentInstance, 'loadPhotos');

    componentInstance.handleSearchChange('test');

    expect(wrapper.state('searchQuery')).toBe('test');
    expect(wrapper.state('page')).toBe(1);
    expect(wrapper.state('loading')).toBe(true);
    expect(spyLoadPhotos.mock.calls.length).toBe(1);
  });
});
