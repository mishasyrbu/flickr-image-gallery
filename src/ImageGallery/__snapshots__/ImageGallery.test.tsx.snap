// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ImageGallery /> it matches the snapshot 1`] = `
<div
  className="image-gallery"
>
  <div
    className="loading-overlay"
  >
    Loading...
  </div>
  <input
    className="search-input"
    onChange={[Function]}
    placeholder="Search..."
    type="text"
    value=""
  />
  <div
    className="results-count"
  >
    Found: 
    0
  </div>
  <div
    className="image-list"
    onScroll={[Function]}
  />
</div>
`;
